### PR TITLE
Initial JREC session recording token support with /jet/jrec route

### PR DESCRIPTION
**Pull request by awakecoding**
- Add new JREC token type for session recording
- Add new jet_rft (recording file type) private claim
- Handle /jet/jrec route for WSS to file streaming
- Improve PowerShell module to generate JREC tokens

### CODE REVIEW

File: devolutions-gateway/src/config.rs [hunk 1]
```diff
@@ -197,7 +203,7 @@ impl Conf {
             sub_provisioner_public_key,
             delegation_private_key,
             plugins: conf_file.plugins.clone(),
-            recording_path: conf_file.recording_path.clone(),
+            recording_path: Some(recording_path),
```
Comment (CBenoit) [inline devolutions-gateway/src/config.rs:206]: With this change, `recording_path` of the `Conf` struct doesn’t need to be an `Option` anymore.

File: devolutions-gateway/src/token.rs [hunk 2]
```diff
@@ -511,6 +536,63 @@ impl<'de> de::Deserialize<'de> for JmuxTokenClaims {
...
 
+// ----- jrec claims ----- //
+
+#[derive(Clone)]
+pub struct JrecTokenClaims {
+    /// Association ID (= Session ID)
+    pub jet_aid: Uuid,
+
+    /// Application protocol
+    pub jet_ap: ApplicationProtocol,
```
Comment (CBenoit) [inline devolutions-gateway/src/token.rs:547]: Leftover from association token?

File: devolutions-gateway/src/token.rs [hunk 3]
```diff
@@ -511,6 +536,63 @@ impl<'de> de::Deserialize<'de> for JmuxTokenClaims {
...
+
+    // Recording file type
+    pub jet_rft: RecordingFileType,
+
+    /// Max session duration
+    pub jet_ttl: SessionTtl,
```
Comment (CBenoit) [inline devolutions-gateway/src/token.rs:553]: Leftover from association token?

File: devolutions-gateway/src/token.rs [hunk 4]
```diff
@@ -511,6 +536,63 @@ impl<'de> de::Deserialize<'de> for JmuxTokenClaims {
...
+            jet_aid: claims.jet_aid,
+            jet_ap: claims.jet_ap,
+            jet_rft: claims.jet_rft,
+            jet_ttl,
+            exp: claims.exp,
+            jti: claims.jti,
+        })
+    }
+}
+
```
Comment (CBenoit) [inline devolutions-gateway/src/token.rs:595]: In this case, you can derive `Deserialize` automatically.

File: powershell/DevolutionsGateway/Public/DGateway.ps1 [hunk 5]
```diff
@@ -832,12 +836,36 @@ function New-DGatewayToken {
...
+
+        $Payload | Add-Member -MemberType NoteProperty -Name 'jet_aid' -Value $AssociationId
+
+        $Payload | Add-Member -MemberType NoteProperty -Name 'dst_hst' -Value $DestinationHost
+    }
+
+    if ($Type -eq 'JREC') {
+        if (-Not $DestinationHost) {
+            throw "DestinationHost is required"
+        }
```
Comment (CBenoit) [inline powershell/DevolutionsGateway/Public/DGateway.ps1:858]: My understanding is that we don’t need this when generating a JREC token.

File: devolutions-gateway/src/token.rs (hunk 4 above)
Comment (awakecoding) [reply devolutions-gateway/src/token.rs]: I'll ask you how to do it next time, it's not much of an issue to have a bit too much boilerplate for the current PR

**Review by thenextman — APPROVED** [review]

**Pull request merged**